perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_group_resource_attribute_def_def_o365EmailAddresses_mu.java

```diff
@@ -0,0 +1,120 @@
+package cz.metacentrum.perun.core.impl.modules.attributes;
+
+import cz.metacentrum.perun.core.api.Attribute;
+import cz.metacentrum.perun.core.api.AttributeDefinition;
+import cz.metacentrum.perun.core.api.AttributesManager;
+import cz.metacentrum.perun.core.api.Group;
+import cz.metacentrum.perun.core.api.Resource;
+import cz.metacentrum.perun.core.api.exceptions.AttributeNotExistsException;
+import cz.metacentrum.perun.core.api.exceptions.GroupResourceMismatchException;
+import cz.metacentrum.perun.core.api.exceptions.InternalErrorException;
+import cz.metacentrum.perun.core.api.exceptions.WrongAttributeAssignmentException;
+import cz.metacentrum.perun.core.api.exceptions.WrongAttributeValueException;
+import cz.metacentrum.perun.core.api.exceptions.WrongReferenceAttributeValueException;
+import cz.metacentrum.perun.core.bl.AttributesManagerBl;
+import cz.metacentrum.perun.core.impl.PerunSessionImpl;
+import cz.metacentrum.perun.core.impl.Utils;
+import cz.metacentrum.perun.core.implApi.modules.attributes.ResourceGroupAttributesModuleAbstract;
+import cz.metacentrum.perun.core.implApi.modules.attributes.ResourceGroupAttributesModuleImplApi;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Matcher;
+
+import static cz.metacentrum.perun.core.impl.Utils.emailPattern;
+import static cz.metacentrum.perun.core.impl.Utils.hasDuplicate;
+
+/**
+ * Module for email addresses for Office365 at Masaryk University.
+ * Implements checks for attribute urn:perun:group-resource:attribute-def:def:o365EmailAddresses_mu.
+ *
+ * Requirements:
+ * <ul>
+ *     <li>type is list</li>
+ *     <li>all values are email addresses</li>
+ *     <li>must contain at least one value if urn:perun:group-resource:attribute-def:def:adName is set</li>
+ *     <li>no duplicates among the list values</li>
+ *     <li>no duplicates among all values of this attribute and all values of
+ *     attribute urn:perun:member:attribute-def:def:o365EmailAddresses:mu</li>
+ * </ul>
+ *
+ * @author Martin Kuba &lt;makub@ics.muni.cz>
+ */
+@SuppressWarnings("unused")
+public class urn_perun_group_resource_attribute_def_def_o365EmailAddresses_mu extends ResourceGroupAttributesModuleAbstract implements ResourceGroupAttributesModuleImplApi {
+
+	private static final String NAMESPACE = AttributesManager.NS_GROUP_RESOURCE_ATTR_DEF;
+	static final String ADNAME_ATTRIBUTE = NAMESPACE + ":adName";
+
+	@Override
+	public void checkAttributeValue(PerunSessionImpl sess, Resource resource, Group group, Attribute attribute) throws InternalErrorException, WrongAttributeValueException, WrongReferenceAttributeValueException, WrongAttributeAssignmentException {
+		List<String> emails;
+		//get values
+		Object value = attribute.getValue();
+		if (value == null) {
+			throw new WrongAttributeValueException(attribute, resource, group, "can't be null.");
```

Comment:
Shouldn't this be checked after check on adName value existence ? What do you think @licehammer ? I'm not sure, what is desired workflow with adName itself - regarding later manual changes and fixes, etc.
Also we translate empty list to null when storing (not sure on retrieving), so you might never get "empty list" on line 73.

Reply:
OK, I will allow null value representing empty list.